### DISCUSSION TITLE
How to set the path to images if they are to be rendered from server side in a golem shiny app?

### DISCUSSION THREAD

**Discussion by udurraniAtPresage** [ThinkR-open/golem#1137] [Q&A] [answered]
I have asked this question on [stackoverflow](https://stackoverflow.com/questions/78275094/how-to-set-the-path-to-images-if-they-are-to-be-rendered-from-server-side-in-a-g) also, but no response for some time now.  

## Goal
In a shiny app, render multiple ggplot plots with images as facet labels.
You can see the complete source code [here](https://github.com/udurraniAtPresage/example).

## Details and Problem
I want to create two plots both of which use the same images. In a normal shiny app, I can read the images in `global.R` and then use them to create plots in the server part.
In the golem framework, I created a `R/global.R` file with this code:

    labels_flowers <- c(
      setosa     = paste0("<img src='", "inst/app/www", "/", "setosa.jpg' width='100'/>"),
      versicolor = paste0("<img src='", "inst/app/www", "/", "versicolor.jpg' width='100'/>"),
      virginica  = paste0("<img src='", "inst/app/www", "/", "virginica.jpg' width='100'/>")
    )

And then I used them in two modules.

### Module example
Consider this module:

    #' plot UI Function
    #'
    #' @description A shiny Module.
    #'
    #' @param id,input,output,session Internal parameters for {shiny}.
    #'
    #' @noRd
    #'
    #' @import shiny
    #' @import ggplot2
    #' @import ggtext
    mod_plot_ui <- function(id){
      ns <- NS(id)
      tagList(
        plotOutput(ns("myplot"))
      )
    }
    
    #' plot Server Functions
    #'
    #' @noRd
    mod_plot_server <- function(id, data){
      moduleServer( id, function(input, output, session){
        ns <- session$ns
        output$myplot <- renderPlot({
          ggplot(data = data()) +
            geom_histogram(aes(Sepal.Length)) +
            facet_wrap(~ Species, labeller = as_labeller(labels_flowers)) +
            theme(strip.text.x = ggtext::element_markdown(size = 20))
        })
      })
    }

The `app_ui.R` and `app_server.R` use the modules as:

#### `app_ui.R`

    #' The application User-Interface
    #'
    #' @param request Internal parameter for `{shiny}`.
    #'     DO NOT REMOVE.
    #' @import shiny
    #' @noRd
    app_ui <- function(request) {
      tagList(
        # Leave this function for adding external resources
        golem_add_external_resources(),
        # Your application UI logic
        fluidPage(
          h1("example"),
          mod_plot_ui("plot1"),
          mod_plot2_ui("plot2")
        )
      )
    }
    
    #' Add external Resources to the Application
    #'
    #' This function is internally used to add external
    #' resources inside the Shiny application.
    #'
    #' @import shiny
    #' @importFrom golem add_resource_path activate_js favicon bundle_resources
    #' @noRd
    golem_add_external_resources <- function() {
      add_resource_path(
        "www",
        app_sys("app/www")
      )
    
      tags$head(
        favicon(),
        bundle_resources(
          path = app_sys("app/www"),
          app_title = "example"
        )
        # Add here other external resources
        # for example, you can add shinyalert::useShinyalert()
      )
    }

#### `app_server.R`

    #' The application server-side
    #'
    #' @param input,output,session Internal parameters for {shiny}.
    #'     DO NOT REMOVE.
    #' @import shiny
    #' @import ggplot2
    #' @import ggtext
    #' @noRd
    app_server <- function(input, output, session) {
    
      df_iris <- reactive(iris)
    
      mod_plot_server("plot1", df_iris)
      mod_plot2_server("plot2", df_iris)
    
    }

## App works locally and on shinyapps but not on Google Cloud Run
App works fine locally and on [shinyapps](https://presage.shinyapps.io/example/). But the [app deployed via Dockerfile on Google Cloud Run](https://example-vlmohgon2q-uc.a.run.app/) throws error:

[![enter image description here][1]][1]

What do I change in the path to images to make sure that the app works fine on Google Cloud Run?

  [1]: https://i.stack.imgur.com/cubOa.png

**Comment by VincentGuyader** [comment, marked as answer]
Hi

You May fin this usefull: https://github.com/VincentGuyader/golimg

:)

**Reply by udurraniAtPresage** [reply]
Thank you.
